src/ingestion/kafka-queue.ts

```diff
@@ -39,27 +37,29 @@ export class KafkaQueue implements Queue {
         isStale,
     }: EachBatchPayload): Promise<void> {
         const batchProcessingTimer = new Date()
-        const rawEvents: RawEventMessage[] = batch.messages.map((message) => ({
-            ...JSON.parse(message.value!.toString()),
-            kafka_offset: message.offset,
-        }))
-        const parsedEvents = rawEvents.map((rawEvent) => ({
-            ...rawEvent,
-            data: JSON.parse(rawEvent.data),
-        }))
-        const pluginEvents: PluginEvent[] = rawEvents.map((rawEvent) => {
-            const { data: dataStr, ...restOfRawEvent } = rawEvent
-            const event = { ...restOfRawEvent, ...JSON.parse(dataStr) }
+
+        const uuidOrder = new Map<string, number>()
+        const uuidOffset = new Map<string, string>()
+        const pluginEvents = batch.messages.map((message, index) => {
+            const { data: dataStr, ...rawEvent } = JSON.parse(message.value!.toString())
+            const event = { ...rawEvent, ...JSON.parse(dataStr) }
+            uuidOrder.set(event.uuid, index)
+            uuidOffset.set(event.uuid, message.offset)
```

Comment:
OK, so this still doesn't address cases like: we've got events `a`, `b`, `c`; `b` and `c` gets discarded; we never resolve `b`'s and `c`'s offsets; in case of restart `b` and `c`, get processed again. But then also it's true that plugins can return whatever the hell they want in any order, so no method can be _completely_ trusted here (though in most cases there won't be shenanigans). Even sorting by UUIDs (which, the ones generated by us, are time-sortable with UUIDT), the `uuid` property may be randomized, so… Alright, whatever. In any case it's better to process an event at-least-once than at-most-once.

Reply:
Yep, this can go deep...